### ISSUE TITLE
[BUG] TypeError: NoneType is not iterable

### ISSUE THREAD

**Issue by mantis** [almenscorner/IntuneCD#159]
When running a backup of the intune configuration, we receive a TypeError: NoneType is not iterable

**Screenshots**
<img width="1162" alt="image" src="https://github.com/almenscorner/IntuneCD/assets/16276382/923619b5-e4dd-45af-b224-4bc0157feb69">

**Run type (please complete the following information):**
Azure Devops Pipeline

**Additional context**

Starting: IntuneCD backup
==============================================================================
Task         : Bash
Description  : Run a Bash script on macOS, Linux, or Windows
Version      : 3.231.5
Author       : Microsoft Corporation
Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/bash
==============================================================================
Generating script.
========================== Starting Command Output ===========================
/usr/bin/bash /home/vsts/work/_temp/5edf2cfa-1763-4171-a40f-4054624eab9b.sh
Backing up App Configuration: NSCT - Defender Endpoint Supervised Policy
Traceback (most recent call last):
  File "/home/vsts/.local/bin/IntuneCD-startbackup", line 8, in <module>
    sys.exit(start())
  File "/home/vsts/.local/lib/python3.10/site-packages/IntuneCD/run_backup.py", line 271, in start
    run_backup(
  File "/home/vsts/.local/lib/python3.10/site-packages/IntuneCD/run_backup.py", line 218, in run_backup
    backup_intune(results, path, output, exclude, token, prefix, append_id, args)
  File "/home/vsts/.local/lib/python3.10/site-packages/IntuneCD/backup_intune.py", line 14, in backup_intune
    results.append(savebackup(path, output, exclude, token, prefix, append_id))
  File "/home/vsts/.local/lib/python3.10/site-packages/IntuneCD/backup/Intune/backup_AppProtection.py", line 49, in savebackup
    assignments = get_object_assignment(profile["id"], assignment_responses)
  File "/home/vsts/.local/lib/python3.10/site-packages/IntuneCD/intunecdlib/graph_batch.py", line 284, in get_object_assignment
    assignments_list = [
TypeError: 'NoneType' object is not iterable

##[error]Bash exited with code '1'.
##[error]Bash wrote one or more lines to the standard error stream.
##[error]Traceback (most recent call last):
  File "/home/vsts/.local/bin/IntuneCD-startbackup", line 8, in <module>

##[error]    sys.exit(start())
  File "/home/vsts/.local/lib/python3.10/site-packages/IntuneCD/run_backup.py", line 271, in start
    run_backup(
  File "/home/vsts/.local/lib/python3.10/site-packages/IntuneCD/run_backup.py", line 218, in run_backup
    backup_intune(results, path, output, exclude, token, prefix, append_id, args)
  File "/home/vsts/.local/lib/python3.10/site-packages/IntuneCD/backup_intune.py", line 14, in backup_intune
    results.append(savebackup(path, output, exclude, token, prefix, append_id))
  File "/home/vsts/.local/lib/python3.10/site-packages/IntuneCD/backup/Intune/backup_AppProtection.py", line 49, in savebackup
    assignments = get_object_assignment(profile["id"], assignment_responses)
  File "/home/vsts/.local/lib/python3.10/site-packages/IntuneCD/intunecdlib/graph_batch.py", line 284, in get_object_assignment
    assignments_list = [
TypeError: 'NoneType' object is not iterable

Finishing: IntuneCD backup

**Comment by almenscorner** [comment]
Hi, can you verify that you do not have any App Protections that has an assignments towards a group that has been removed?

**Comment by mantis** [comment]
If i'm looking in the right place... I don't think we've ever used this bit (I'm assuming what's on screen is something Microsoft added by default):

![image](https://github.com/almenscorner/IntuneCD/assets/16276382/e414c10f-4a72-4f53-be3a-d5e1b6c2bcac)

![image](https://github.com/almenscorner/IntuneCD/assets/16276382/8120f170-3c1d-4608-9527-8c78401f2fd4)

![image](https://github.com/almenscorner/IntuneCD/assets/16276382/30c30f5e-7775-4cc8-873c-aa45e32b0824)

**Comment by almenscorner** [comment]
I don't think MS has added any App Protection policy by default, at least I do not have one like that in my env. If you are not using it, can you remove it and try a new backup?

**Comment by mantis** [comment]
@almenscorner I've not done anything yet, as I wanted to understand where this policy has come from - I think I've now worked it out:

https://learn.microsoft.com/en-us/mem/intune/apps/app-protection-policy

There is a section "App protection global policy":

```
If a OneDrive administrator browses to admin.onedrive.com and selects Device access, they can set Mobile application management controls to the OneDrive and SharePoint client apps.

The settings, made available to the OneDrive Admin console, configure a special Intune app protection policy called the Global policy. This global policy applies to all users in your tenant, and has no way to control the policy targeting.

Once enabled, the OneDrive and SharePoint apps for iOS/iPadOS and Android are protected with the selected settings by default. An IT Pro can edit this policy in the [Microsoft Intune admin center](https://go.microsoft.com/fwlink/?linkid=2109431) to add more targeted apps and to modify any policy setting.

By default, there can only be one Global policy per tenant. However, you can use [Intune Graph APIs](https://learn.microsoft.com/en-us/mem/intune/developer/intune-graph-apis) to create extra global policies per tenant, but doing so isn't recommended. Creating extra global policies isn't recommended because troubleshooting the implementation of such a policy can become complicated.

While the Global policy applies to all users in your tenant, any standard Intune app protection policy will override these settings.
```

There is then a note: "The policy settings in the OneDrive Admin Center are no longer being updated. Microsoft Intune may be used instead".

So looking at that, I think we probably need to work out a code fix for this as presumably there will be other users that have a non-targetted policy here...

**Comment by almenscorner** [comment]
Looking at that, we could always skip assignment if this is the odata type `"@odata.type": "#microsoft.graph.defaultManagedAppProtection"`

**Comment by mantis** [comment]
I'd say that would seem the most reasonable fix - Microsoft seem to suggest there will only ever be one of these policies - with the disclaimer that you could technically make your own

**Comment by almenscorner** [comment]
I've included that in 2.0.8 beta, install the beta and let me know if the backup now is successful `pip install IntuneCD==2.0.8b1`

**Comment by mantis** [comment]
updated our pipeline - seems to build again  so this is fixed as of 2.0.8 and tissue can be closed :)

![image](https://github.com/almenscorner/IntuneCD/assets/16276382/ab392265-373d-40cc-8e66-74be14594982)

**Comment by almenscorner** [comment]
Cool, I'll merge the changes to main and close this issue! :)
